### DISCUSSION TITLE
Plotting bootstrap values on a tree using a file with bootstrap replicates

### DISCUSSION THREAD

**Discussion by veronicamb** [iqtree/iqtree2#305] [Q&A] [answered]
Hello, 

I have generated 200 bootstrap replicates using an array job and have combined them all into a single file. However, I'm now struggling to use this file to plot the bootstrap proportions onto my main tree. Is there a way to do this in IQ-TREE?

I have tried a couple of different commands, but I keep getting the same error message:
ERROR: You must specify the number of taxa in the PD set.

I don't understand what PD is, or how I should inform the program on the number of taxa.

thank you!

Veronica

**Comment by veronicamb** [comment]
I figured it out!

**Reply by roblanf** [reply]
Awesome! Any chance you could post your solution here (maybe all the commands, i.e. generate the ML tree, generate the bootstrap trees, map them back). This would help anyone who lands here from google with a similar problem.

**Reply by veronicamb** [reply, marked as answer]
Of course! My bad, I should have done that before closing the thread. Here it is:

I'm working on a large dataset (~1000 taxa), which makes running bootstraps time-consuming. The best solution was to run each bootstrap replicate in parallel (using many computer nodes) to speed up the analysis.

I first ran a partitioned ML analysis with ultrafast bootstraps.

```shell
iqtree -s datamatrix.nex -spp partitions.nex -m MFP+MERGE -rcluster 30 -pre ML_tree -nt AUTO -bb 1000
```

Then I ran 200 nonparametric bootstraps using an array script (set up in the SLURM header)

```shell
iqtree -s datamatrix.nex -spp ML_tree.best.scheme.nex -pre bootstrap_replicates/bootstrap_$SLURM_ARRAY_TASK_ID --nt AUTO -b 1
```

After the run, the directory "bootstrap_replicates" contains 200 bootstrap_*.treefile files (bootstrap_1.treefile, bootstrap_2.treefile, bootstrap_3.treefile, ..., bootstrap_200.treefile), which I concatenated into a single file.

```shell
cat bootstrap_replicates/bootstrap_*.treefile > 200_boostraps.treefile
```

Then, in iqtree, I used this file to calculate bootstrap proportions and map them on my ML tree

```shell
iqtree -t 200_bootstraps.treefile -sup ML_tree.treefile -pre ML_tree_sup -nt AUTO
```

* `-t`: file with bootstrap replicates
* `-sup`: target tree, on which the proportions should be mapped

iqtree appended the bootstrap proportions to the ultrafast bootstrap values that were already on the tree.

**Reply by roblanf** [reply]
Such an excellent and detailed explanation! Thanks!

**Reply by roblanf** [reply]
I am coming back to this simply because I found it through google myself, and I wanted to add that there's a simple (non-SLURM) way of running the bootstrap replicates, which looks like this:


```bash
seq 1 1000 | parallel --bar --eta -j 100 'iqtree -s aln.phy -b 1 -seed {} -T 1 -pre rep_{}`
```

Breakdown:

* `seq 1 1000` runs from 1 to 1000 (change the latter number to do different numbers of bootstraps
* `parallel --bar --eta -j 100` uses GNU parallel to run 100 simultaneous jobs, with a progress bar and an ETA
* `-seed {}` uses the number of the rep as the seed (so every bootstrap has a different seed)
* `-pre rep_{}` names files by the number of the rep
* `-T 1` number of threads per bootstrap

Note that the total number of threads used will be whatever you tell GNU parallel multiplied by whatever you tell IQ-TREE. Typically the fastest you can do it (RAM permitting, of course) is to give IQ-TREE 1 thread per bootstrap and give as many as possible to GNU parallel.
